CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

IntegralProgrammer opened the pull request:
This PR implements CARDS-1424 by providing a scheduled task that checks if any patients have any appointments on the third day from the day that the task executes. For example, if the task executes on _2022-01-01_ at _01:30:00_ it will check for any appointments happening at any time on _2022-01-04_. For each appointment that is found, the task will check if the patient has provided their email address and has consented to email communications. If they have, they will receive an email notifying them of their upcoming appointment and prompting them to complete their surveys.

To test:

- Follow the instructions at https://github.com/data-team-uhn/cards/pull/822#issue-1063852994 to configure a local debugging mail server for CARDS to use.
- Start CARDS with the command `CARDS_HOST_AND_PORT='datapro.uhn.ca' CLINIC_SLING_PATH='/Proms.html/Cardio' PATIENT_NOTIFICATION_FROM_ADDRESS='datapro@uhn.ca' PATIENT_NOTIFICATION_FROM_NAME='UHN DATAPRO' NIGHTLY_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' SLING_COMMONS_CRYPTO_PASSWORD=password ./start_cards.sh --project cards4proms --dev -f mvn:io.uhndata.cards/cards-email-notifications/0.9-SNAPSHOT/slingosgifeature -V emailnotifications.smtps.port=8465 -V emailnotifications.smtps.host=localhost`.
- Login to CARDS as `admin`:`admin`.
- Create a new _Patient information_ Form and, with it, a new _Patient_ subject (`P1`).
- Complete the _Patient information_ and provide an email address and select `Yes` for the question `The patient consents to be contacted by email`.
- Create a new _Visit information_ Form and, with it, a new _Visit_ subject (`V1`). Select `P1` as the _parent Patient_ for this `V1` subject.
- Set the _Time_ for this Form to be in three days from now. This step will not work in Firefox but will work in Chrome.
- Save the Form.
- Open Composum at `http://localhost:8080/bin/browser.html`
- Make note of the `jcr:uuid` value for the `/Questionnaires/Visit information/surveys` node.
- Find the JCR node under `/Forms/` that has a `question` value equal to the `jcr:uuid` value from the previous step.
- Add a _string_ property named `value` with the value of `Cardio` - here, we are setting the value _Cardio_ as the response to the _Surveys_ question for this _Visit information_ form.
- In the console with Python's SMTP debugging server, you should see email notifications being received. The rate at which the notifications are received is set by the `NIGHTLY_NOTIFICATIONS_SCHEDULE` environment variable which, for testing purposes, I have provided instructions so that it runs at the start of every minute.
- Ensure that the link works and visiting it (replacing `datapro.uhn.ca` with `localhost:8080` and `https` with `http`) in a browser window that is not logged into CARDS works as expected.
- Experiment with different patients with different scheduled visits at different clinics and ensure that everything works as expected.

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Remove @SuppressWarnings for checkstyle:MultipleStringLiterals in the InitialNotificationsTask class

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Rename options to initialNotificationsOptions in preparation for reminderNotificationsOptions to be used in CARDS-1422

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Rename getQuestionAnswerForPatientSubject to getQuestionAnswerForSubject as it works for all Subject types, not just patients

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Remove warning messages used for debugging

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

No need to change start-order in feature.json from 25 to 26

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Replace hard-coded FROM_ADDRESS, FROM_NAME, and SUBJECT email parameters with environment variables

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Link sent to patients to complete their surveys should use the CARDS_HOST_AND_PORT environment variable (introduced in CARDS-1374) and the CLINIC_SLING_PATH environment variable to build the surveys link

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Simplify long comment

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Add Javadoc for the AppointmentUtils class

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Rename patientTypeUuid to subjectTypeUuid as the Subject is not necessarily a Patient - it could be a Visit

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

72 hour notification email should be generated by the rendering of a template fetched from JCR for the appointment's clinic

TODO: Rich-text HTML emails should be sent with the plaintext version only for fallback purposes. This will be implemented once the jakarta.activation.UnsupportedDataTypeException issue is fixed

marta- commented on modules/email-notifications/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+
+Make the most of your appointment by answering your Pre-Appointment Questions.
+
+Answer your Pre-Appointment Questions
```
```suggestion
Answer your Pre-Appointment Questions: ${surveysLink}
```

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Apply 72h.txt template change suggestion from @marta-

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

IntegralProgrammer committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

Email notification subject lines are now defined in the notification tasks and not as a single environment variable as the subject line may differ between notification tasks

acrowthe committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys
Move PROMs specific code into PROMs backend module

sdumitriu commented on modules/email-notifications/pom.xml:
```diff
@@ -56,6 +56,11 @@
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-text</artifactId>
+      <version>1.6</version>
```
The version is already defined in the root pom:

```suggestion
```

sdumitriu commented on proms-resources/backend/pom.xml:
```diff
@@ -32,20 +32,53 @@
         <extensions>true</extensions>
+        <configuration>
+          <instructions>
+            <Sling-Initial-Content>SLING-INF/content/;path:=/</Sling-Initial-Content>
```
```suggestion
            <Sling-Initial-Content>SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates;path:=/apps/cards/proms/clinics/Cardio/mailTemplates;overwrite:=true</Sling-Initial-Content>
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/InitialNotificationsTask.java:
```diff
@@ -0,0 +1,140 @@
+         */
+        final Date today = new Date();
+        final Calendar dateToQuery = Calendar.getInstance();
+        dateToQuery.setTime(today);
```
Why not just use `Calendar.getInstance()`? That also returns a calendar object set to "right now". Is there some timezone trickery that requires it?

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+
+Please note that your responses may not be reviewed by your care team until the day of your appointment. If your symptoms are worsening while waiting for your next appointment, please proceed to your nearest Emergency Department today, or call 911.
+
+This link will expire the day after your appointment. For questions about the details of your appointment, please call the Peter Munk Cardiac Centre Info Line</span> at 416-340-5309 between 8:00 am and 4:30 pm, Monday to Friday, to speak with an Information Specialist.
```
> This link will expire the day after your appointment.

That's not entirely true. Right now the link expires 2 hours after the appointment, so we can either say "This link will expire after your appointment", or change the token to expire not on `visit time + 2h` but on `vist time + 1day @ midnight`. I would just change the email wording to "after the appointment".

marta- replied:
To be fair, I've seen wait times significantly over 2 hours even with appointment. Maybe expiring at midnight isn't such a terrible idea.

sdumitriu replied:
OK, moved to midnight.

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+You are receiving this email because you have an upcoming appointment.
```
Should we provide more details about the appointment (time/place/doctor)? I think it's not needed, they should already receive this info in another email from UHN.

marta- replied:
Not needed I hope

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+
+Instructions
+
+1. Visit ${surveysLink} and log-in to DATAPRO up to two days before your appointment
```
Where does `up to two days before` come from? Can we remove that part?

marta- replied:
I assume it's coming from the fact that we're generating emails ~72h before appointment. So basically if you have an appointment on Thursday, we push emails on Monday night (at a time when you're likely asleep) and you wake up to an email on Tuesday, 2 days before your appointment. So you're able to use the link and provide responses as early as two days before the appointment.

We didn't come up with the wording, this was in the specs.

sdumitriu replied:
Well, they can click on the link right when they receive the email.

sdumitriu replied:
And Monday night is actually Tuesday, so 72h is three days.

k-pace replied:
Let's rephrase: ""... and log-in to DATAPRO. Your link is valid until two hours after your appointment start time."

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+
+3. Make the most of your appointment by bringing up any concerns you have with your provider
+
+4. Note: If your appointment is cancelled or rescheduled and you've already submitted your responses, they will not be reviewed until the time of your appointment. Make sure to contact your nurse triage line, proceed to the ED or call 911 if your symptoms are severe or worsening.
```
"they will not be reviewed until the time of your appointment" is not clear to me. What does that mean?

marta- replied:
*of your new appointment. If you're saying in the responses that you're feeling really down and depressed, maybe nobody will notice until 1 week has passed and you have a new appointment. So don't wait for the appointment, get help now.

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+
+3. Make the most of your appointment by bringing up any concerns you have with your provider
+
+4. Note: If your appointment is cancelled or rescheduled and you've already submitted your responses, they will not be reviewed until the time of your appointment. Make sure to contact your nurse triage line, proceed to the ED or call 911 if your symptoms are severe or worsening.
```
"proceed to the ED" Can we say "Emergency Department" instead?

marta- replied:
okay

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+- UHN is not responsible for the security of patients' internet service providers, email domains, personal devices, personal computers, applications on those devices, etc. We suggest you take all steps recommended by your various providers to maximize the security of those tools.
+- We suggest that you use your personal email when communicating with health care providers such as UHN and not a work email address, as your employer may have the right to inspect your work email.
+- Use a secure internet connection. Do not use a public internet connection such as at the airport, internet cafe, public library or other open public area, where it may be easier to "hack" into your email account and access your personal information.
+- For more information on how to protect yourself against security risks associated with electronic communication, read this article on UHN.ca: Don't let a phishing attack reel you in.
```
"read this article" A link would be useful.

marta- replied:
+1

sdumitriu commented on proms-resources/backend/src/main/resources/SLING-INF/content/apps/cards/proms/clinics/Cardio/mailTemplates/72h.txt:
```diff
@@ -0,0 +1,43 @@
+- Use a secure internet connection. Do not use a public internet connection such as at the airport, internet cafe, public library or other open public area, where it may be easier to "hack" into your email account and access your personal information.
+- For more information on how to protect yourself against security risks associated with electronic communication, read this article on UHN.ca: Don't let a phishing attack reel you in.
+
+Please contact your care provider or the UHN Privacy Office at (416) 340-4800 ext. 6937 if you do not wish to receive email communication from UHN. You may withdraw your consent to receive email communication from UHN at any time.
```
Above we used a different format for the phone number, should we be consistent?

marta- replied:
yes

sdumitriu committed:
[misc] Add support for specifying both "-P project" and "-P cards4project" for the start script

sdumitriu committed:
[misc] Since the cards4proms project requires the email plugin, automatically enable it when specifying this project at startup

sdumitriu committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys
Specify default values for the environment variables

sdumitriu committed:
[cleanup] Remove debug logs, better name for a variable, improve code

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/NightlyNotifications.java:
```diff
@@ -0,0 +1,77 @@
+    protected void activate(ComponentContext componentContext) throws Exception
+    {
+        LOGGER.info("NightlyNotifications activating");
+        final String nightlyNotificationsSchedule = System.getenv("NIGHTLY_NOTIFICATIONS_SCHEDULE");
```
This should have a good default value.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/InitialNotificationsTask.java:
```diff
@@ -0,0 +1,140 @@
+    /** Default log. */
+    private static final Logger LOGGER = LoggerFactory.getLogger(InitialNotificationsTask.class);
+
+    private static final String CARDS_HOST_AND_PORT = System.getenv("CARDS_HOST_AND_PORT");
```
This should have a good default value.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/InitialNotificationsTask.java:
```diff
@@ -0,0 +1,140 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(InitialNotificationsTask.class);
+
+    private static final String CARDS_HOST_AND_PORT = System.getenv("CARDS_HOST_AND_PORT");
+    private static final String CLINIC_SLING_PATH = System.getenv("CLINIC_SLING_PATH");
```
This can now be hardcoded to `/Proms.html` since we no longer require the questionnaire set in the URL.

sdumitriu committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys
Move the token expiration to midnigh, not just 2h after the appointment

sdumitriu committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys
Minor improvements to the email templates

sdumitriu committed:
CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys
Add link to the mentioned article
